File: Packages/MeterKit/Sources/MeterFeatures/Developer/DeveloperDebugLog.swift
#if DEBUG
import Foundation
import os
import MeterProviders

/// 真机调试用的内存日志。os.Logger 在没连 Xcode 时不好拷，这里给「复制全部」。
enum DeveloperDebugLog: Sendable {
    static let limit = 300

    private static let storage = OSAllocatedUnfairLock(initialState: [Line]())

    struct Line: Sendable {
        var at: Date
        var category: String
        var message: String
    }

    static func record(category: String, _ message: String) {
        let line = Line(at: Date(), category: category, message: message)
        storage.withLock { buffer in
            buffer.append(line)
            if buffer.count > limit {
                buffer.removeFirst(buffer.count - limit)
            }
        }
    }

    static var lines: [Line] {
        storage.withLock { $0 }
    }

    static func clear() {
        storage.withLock { $0 = [] }
    }

    @MainActor
    static func exportText(
        dashboard: DashboardModel,
        exchanges: [HTTPExchange]
    ) -> String {
        var parts: [String] = [
            "# TollCat debug log",
            "copiedAt \(ISO8601DateFormatter().string(from: Date()))",
        ]
        if !dashboard.lastRefreshSummary.isEmpty {
            parts.append("lastRefresh \(dashboard.lastRefreshSummary)")
        }
        let events = lines
        if !events.isEmpty {
            parts.append("")
            parts.append("## events")
            let formatter = ISO8601DateFormatter()
            for line in events {
                parts.append("\(formatter.string(from: line.at)) [\(line.category)] \(line.message)")
            }
        }
        if !exchanges.isEmpty {
            parts.append("")
            parts.append("## http")
            for exchange in exchanges {
                parts.append(exchange.summary)
                if !exchange.body.isEmpty {
                    parts.append(exchange.body)
                }
                parts.append("")
            }
        }
        parts.append("## store")
        parts.append(DeveloperStoreDump.json(from: dashboard))
        return parts.joined(separator: "\n")
    }
}
#endif

File: Android/app/src/main/kotlin/com/zhechengqi/tollcat/developer/DeveloperDebugLog.kt
package com.zhechengqi.tollcat.developer

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object DeveloperDebugLog {
    const val LIMIT = 300

    data class Line(
        val atMillis: Long,
        val category: String,
        val message: String,
    )

    data class HttpExchange(
        val summary: String,
        val body: String,
    )

    private val events = ArrayDeque<Line>()
    private val http = ArrayDeque<HttpExchange>()

    @Synchronized
    fun record(category: String, message: String) {
        events.addLast(Line(System.currentTimeMillis(), category, message))
        while (events.size > LIMIT) events.removeFirst()
    }

    @Synchronized
    fun recordHttp(summary: String, body: String) {
        http.addLast(HttpExchange(summary, body))
        while (http.size > LIMIT) http.removeFirst()
    }

    @Synchronized
    fun lines(): List<Line> = events.toList()

    @Synchronized
    fun exchanges(): List<HttpExchange> = http.toList()

    @Synchronized
    fun clear() {
        events.clear()
        http.clear()
    }

    fun exportText(storeDump: String, lastRefresh: String?): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        val parts = mutableListOf(
            "# TollCat debug log",
            "copiedAt ${formatter.format(Date())}",
        )
        if (!lastRefresh.isNullOrBlank()) {
            parts += "lastRefresh $lastRefresh"
        }
        val events = lines()
        if (events.isNotEmpty()) {
            parts += ""
            parts += "## events"
            for (line in events) {
                parts += "${formatter.format(Date(line.atMillis))} [${line.category}] ${line.message}"
            }
        }
        val exchanges = exchanges()
        if (exchanges.isNotEmpty()) {
            parts += ""
            parts += "## http"
            for (exchange in exchanges) {
                parts += exchange.summary
                if (exchange.body.isNotBlank()) {
                    parts += exchange.body
                }
                parts += ""
            }
        }
        parts += "## store"
        parts += storeDump
        return parts.joinToString("\n")
    }
}
